rbd.c

```diff
@@ -239,6 +244,54 @@ static void tcmu_rbd_service_status_update(struct tcmu_device *dev,
 
 #endif /* LIBRADOS_SUPPORTS_SERVICES */
 
+static int tcmu_rm_stale_entries_from_blacklist(struct tcmu_device *dev)
+{
+	struct tcmu_rbd_state *state = tcmur_dev_get_private(dev);
+	char *cmd = NULL;
+	char **entry, *addr;
+	int ret = 0;
+	int i = 0, j;
+
+	pthread_mutex_lock(&blacklist_caches_lock);
+	if (darray_empty(blacklist_caches))
+		goto unlock;
+
+	/* Try to remove all the stale blacklist entities */
+	darray_foreach(entry, blacklist_caches) {
+		tcmu_dev_info(dev, "removing addrs: {%s}\n", *entry);
+		ret = asprintf(&cmd,
+			      "{\"prefix\": \"osd blacklist\","
+			      "\"blacklistop\": \"rm\","
+			      "\"addr\": \"%s\"}", *entry);
```

Comment:
This probably needs to be tested. That `rados_getaddrs` call might return a address vector (i.e. `[v2:192.168.1.3:40945,v1:192.168.1.3:40946]`) but blacklisting might only act on a single address (i.e. `192.168.1.3:0/<nonce>`).

Reply:
Yeah, make sense, I will test this.

Reply:
Currently in none `DPDK` case, there will only one `processer` in the `Messenger `, and then the `Messenger::my_addrs` will only be single one, like the logs:

```
 124 2020-08-05T14:30:18.410+0800 7f579b7fe700 25 --  >> [v2:10.72.47.137:40500/0,v1:10.72.47.137:40501/0] conn(0x55726e5fb930 msgr2=0x557     26e60bca0 unknown :-1 s=STATE_CONNECTION_ESTABLISHED l=1)._try_send cs.send 72 bytes
 125 2020-08-05T14:30:18.410+0800 7f579bfff700 10 -- 10.72.47.136:0/3287216816 learned_addr had no addrs                                  
 126 2020-08-05T14:30:18.410+0800 7f579bfff700  1 -- 10.72.47.136:0/3287216816 learned_addr learned my addr 10.72.47.136:0/3287216816 (peer_addr_for_me v2:10.72.47.137:0/0)
```

It will be `learned my addr 10.72.47.136:0/3287216816`, not the `peer_addr_for_me v2:10.72.47.137:0/0`

The `[v2:192.168.1.3:40945,v1:192.168.1.3:40946]` you mentioned above you meant the monitor listening addresses in `ceph.conf`, right ? If so, even if the cluster enables msgr1 and msgr2 at the same time, the monclient will only connect to one of them and it will always choose the first one from `[v2:XXXX,v1:XXXX]`.

We won't use the DPDK here, right ?

Reply:
> The `[v2:192.168.1.3:40945,v1:192.168.1.3:40946]` you mentioned above you meant the monitor listening addresses in `ceph.conf`, right ? If so, even if the cluster enables msgr1 and msgr2 at the same time, the monclient will only connect to one of them and it will always choose the first one from `[v2:XXXX,v1:XXXX]`.

It's still a `entity_addrvec_t` internally which is converted to a string. I just want to make sure the result of that API call matches the `10.72.47.136:0/3287216816`-style format.
 
> We won't use the DPDK here, right ?

It won't be using DPDK. It would be using the standard POSIX async messenger.

Reply:
> It's still a entity_addrvec_t internally which is converted to a string. I just want to make sure the result of that API call matches the 10.72.47.136:0/3287216816-style format.

Yeah, that make sense.

Checked it code, when there has only one address in that vector it will return a string, it will be like `10.72.47.136:0/3287216816`, only when there has more than one addresses, it will be `['10.72.47.136:0/3287216816', '10.72.47.136:0/4587216543']`.

The code:

```
708   friend std::ostream& operator<<(std::ostream& out, const entity_addrvec_t& av) {
709     if (av.v.empty()) {
710       return out;
711     } else if (av.v.size() == 1) {
712       return out << av.v[0];
713     } else {                                                                                                                          
714       return out << av.v;
715     }
716   }
```

But it will be always wise to compatible to this case, and I will fix it.
Thanks